pulsar/src/test/java/com/hazelcast/jet/contrib/pulsar/PulsarSourceTest.java

```diff
@@ -0,0 +1,264 @@
+import com.hazelcast.jet.pipeline.StreamSource;
+import com.hazelcast.jet.pipeline.test.AssertionCompletedException;
+
+import com.hazelcast.jet.pipeline.test.AssertionSinks;
+import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.client.api.PulsarClientException;
+import org.apache.pulsar.client.api.Schema;
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletionException;
+
+import static com.hazelcast.jet.core.test.JetAssert.assertEquals;
+import static com.hazelcast.jet.core.test.JetAssert.assertTrue;
+import static com.hazelcast.jet.core.test.JetAssert.fail;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+
+public class PulsarSourceTest extends PulsarTestSupport {
+    private static final int ITEM_COUNT = 1_000;
+    private JetInstance jet;
+
+    @Before
+    public void setup() {
+        jet = createJetMember();
+    }
+
+    @After
+    public void after() {
+        jet.shutdown();
+
+    }
+
+    @AfterClass
+    public static void afterClass() throws PulsarClientException {
+        PulsarTestSupport.shutdown();
+    }
+
+    @Test
+    public void when_projectionFunctionProvided_thenAppliedToReadRecords() {
+        String topicName = randomName();
+        Map<String, Object> consumerConfig = new HashMap<>();
+        consumerConfig.put("consumerName", "hazelcast-jet-consumer");
+        consumerConfig.put("subscriptionName", "hazelcast-jet-subscription");
+
+        StreamSource<String> pulsarTestStream = PulsarSources.pulsarConsumer(
+                Collections.singletonList(topicName),
+                2,
+                consumerConfig,
+                () -> PulsarClient.builder().serviceUrl(PulsarTestSupport.getServiceUrl()).build(),
+                () -> Schema.BYTES,
+                x -> new String(x.getData()) + "-suffix");
+        Pipeline pipeline = Pipeline.create();
+        pipeline.readFrom(pulsarTestStream)
+                .withoutTimestamps()
+                .writeTo(AssertionSinks.assertCollectedEventually(60,
+                        list -> {
+                            assertEquals("# of Emitted items should be equal to # of published items",
+                                    ITEM_COUNT, list.size());
+                            for (int i = 0; i < ITEM_COUNT; i++) {
+                                String message = "hello-pulsar-" + i + "-suffix";
+                                Assert.assertTrue("missing entry: " + message, list.contains(message));
+                            }
+                        })
+                );
+        Job job = jet.newJob(pipeline);
+        sleepAtLeastSeconds(5);
```

Comment:
What is this waiting for exactly? Waiting exact period usually leads to a flaky test.